humility-core/src/arch/mod.rs

```diff
@@ -0,0 +1,159 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::arch::arm::ARMArch;
+use crate::arch::rv::RVArch;
+use crate::hubris::{HubrisArchive, HubrisStruct, HubrisTarget};
+use crate::regs::Register;
+use anyhow::{bail, Result};
+use capstone::prelude::*;
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+
+pub mod arm;
+pub mod rv;
+
+pub trait Arch {
+    ///
+    /// Return elf header `e_machine` for this architecture
+    ///
+    fn get_e_machine(&self) -> u16;
+
+    ///
+    /// Return elf header byte `EI_CLASS` from `e_ident` for this architecture
+    ///
+    fn get_ei_class(&self) -> u8;
+
+    ///
+    /// Returns the instruction used to trigger a syscall
+    ///
+    fn get_syscall_insn(&self) -> u32;
+    ///
+    /// Return the register used as the return addr for function callsj
+    ///
+    fn get_ret_reg(&self) -> Register;
+    ///
+    /// Return the stack pointer register
+    ///
+    fn get_sp(&self) -> Register;
+    ///
+    /// Return the pc register
+    ///
+    fn get_pc(&self) -> Register;
+
+    ///
+    /// Returns all the general purpose/ integer registers
+    ///
+    fn get_all_gpr(&self) -> Vec<Register>;
+
+    ///
+    /// Return all the registers Humlility supports
+    ///
+    fn get_all_registers(&self) -> Vec<Register>;
+
+    ///
+    /// Convert a dwarf id into a Register object
+    ///
+    fn register_from_dwarf_id(&self, id: u32) -> Result<Register>;
+
+    fn register_from_id(&self, id: u32) -> Result<Register>;
+
+    fn get_syscall_register(&self, arg_number: u8) -> Result<Register>;
+
+    fn get_generic_chip(&self) -> String;
+
+    fn presyscall_pushes(
+        &self,
+        cs: &Capstone,
+        instrs: &[capstone::Insn],
+    ) -> Result<Vec<Register>>;
+
+    fn read_saved_task_regs(
+        &self,
+        regs: &[u8],
+        state: &HubrisStruct,
+        hubris: &HubrisArchive,
+        core: &mut dyn crate::core::Core,
+    ) -> Result<BTreeMap<Register, u32>>;
+
+    fn make_capstone(&self) -> Result<Capstone>;
+
+    ///
+    /// will clear the thumb bit, but only if the current architecture is arm
+    ///
+    fn try_clear_thumb(&self, data: u32) -> u32 {
```

Comment:
I think we should eventually remove any references to thumb in common code.

Reply:
I would agree, but I could not think of a good way to do it.  It is only used in 2 places, but is embedded into code that is otherwise arch agnostic.  Any suggestions?

Reply:
Yeah, it seems pretty embedded in the arch agnostic code. That is why it also feels very wrong. I don't have a great suggestion. But one way to handle this could be to use a more generic function name. From the context, it looks like this function is called on instruction addresses to store value without the thumb bit. You can probably name it `extract_instr_addr()` or `get_instr_addr()` and provide a comment saying that the function can be used by arch-specific code to post-process any address value to get the instruction address by dropping any arch-specific mode bits. e.g. On ARM, it can be used to return address without the Thumb bit. For architectures that do not store any mode bits in the address, there is no need to implement the trait function.

Reply:
Sounds good.  I went ahead and renamed `get_fn_pointer`.